discord/http.py

```diff
@@ -349,159 +349,170 @@ class Ratelimit:
-        exception = RateLimited(self.reset_after) if error else None
-        async with self._sleeping:
-            if not error:
-                await asyncio.sleep(self.reset_after)
+        # 3. Always use the lowest remaining value
+        if remaining < self.remaining:
+            self.remaining = remaining
+            update = True
 
-        self.reset()
-        self._wake(self.remaining, exception=exception)
+        self.limit = limit
+        self.one_shot = False
 
-    def is_expired(self) -> bool:
-        return self.expires is not None and self._loop.time() > self.expires
+        # Return whether this update was relevant or not.
+        return update
 
     def is_inactive(self) -> bool:
-        delta = self._loop.time() - self._last_request
-        return delta >= 300 and self.outgoing == 0 and len(self._pending_requests) == 0
-
-    async def acquire(self) -> None:
-        self._last_request = self._loop.time()
-        if self.is_expired():
-            self.reset()
-
-        if self._max_ratelimit_timeout is not None and self.expires is not None:
-            # Check if we can pre-emptively block this request for having too large of a timeout
-            current_reset_after = self.expires - self._loop.time()
-            if current_reset_after > self._max_ratelimit_timeout:
-                raise RateLimited(current_reset_after)
-
-        while self.remaining <= 0:
-            future = self._loop.create_future()
-            self._pending_requests.append(future)
-            try:
-                while not future.done():
-                    # 30 matches the smallest allowed max_ratelimit_timeout
-                    max_wait_time = self.expires - self._loop.time() if self.expires else 30
-                    await asyncio.wait([future], timeout=max_wait_time)
-                    if not future.done():
-                        await self._refresh()
-            except:
-                future.cancel()
-                if self.remaining > 0 and not future.cancelled():
-                    self._wake_next()
-                raise
-
-        self.remaining -= 1
-        self.outgoing += 1
+        delta = self.http.loop.time() - self._last_request
+        return delta >= 300 and (self.one_shot or (self.outgoing == 0 and self.pending == 0))
+
+    def _wake(self) -> None:
+        if self._future and not self._future.done():
+            self._future.set_result(None)
+
+    async def _wait_global(self, start_time: float):
+        # Sleep up to 3 times, to account for global reset at overwriting during sleeps
+        for i in range(3):
+            seconds = self.http.global_reset_at - start_time
+            if seconds > 0:
+                await asyncio.sleep(seconds)
+                continue
+            break
+        else:
+            raise RuntimeError('Global reset at changed more than 3 times')
+
+    async def _wait(self):
+        # Consider waiting if none is remaining
+        if not self.remaining:
```

Comment:
I suggest using explicit checks instead of passive boolean ones for all the integer comparisons.
```suggestion
        if self.remaining == 0:
```